distributed/worker.py

```diff
@@ -3044,7 +3046,7 @@ async def gather_dep(
                 for d in has_what:
                     ts = self.tasks[d]
                     ts.who_has.remove(worker)
-                    if not ts.who_has:
+                    if not ts.who_has and ts.state not in ("released", "memory"):
```

Comment:
I see the failure in https://github.com/dask/distributed/runs/6032728170?check_suite_focus=true
```python
            assert all(
                dep.key in self.data or dep.key in self.actors
                for dep in ts.dependencies
            )
```
but I'm not following why relaxing this from `ts.state in ("fetch", "flight")` caused it. I accept that it seems to work, but I'd like to understand it.

If you'd be willing to add an `assert ts.state in ("fetch", "flight"), ts.state` within this current conditional, I'd be very curious to see what fails in CI. I can't really make sense of why we'd want to transition tasks in any other state to missing.

Reply:
> I'd like to understand it.

I encourage you to understand it :) 

> If you'd be willing to add an assert ts.state in ("fetch", "flight"), ts.state within this current conditional, I'd be very curious to see what fails in CI

I think that we should do this in another PR and get this in as it is.  I'd like to resolve this deadlock and get out a release.